bigtable/client/internal/endian.h

```diff
@@ -0,0 +1,150 @@
+// Copyright 2018 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_INTERNAL_ENDIAN_H_
+#define GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_INTERNAL_ENDIAN_H_
+
+#include <bitset>
+#include <limits>
+#include <vector>
+
+#include "bigtable/client/internal/strong_type.h"
+#include "bigtable/client/internal/throw_delegate.h"
+
+#ifdef _MSC_VER
+#include <stdlib.h>
+#elif defined(__GNUC__) || defined(__clang__)
+#include <byteswap.h>
+#endif
+
+#include <string>
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+using bigendian64_t = internal::StrongType<std::int64_t, struct BigEndianType>;
+
+namespace internal {
+
+#ifndef _MSC_VER
+#ifndef __GNUC__
+#ifndef __clang__
+static std::int64_t explicit_swap64(std::int64_t value) {
+  std::int64_t return_value = ((((value)&0xff00000000000000ull) >> 56) |
+                               (((value)&0x00ff000000000000ull) >> 40) |
+                               (((value)&0x0000ff0000000000ull) >> 24) |
+                               (((value)&0x000000ff00000000ull) >> 8) |
+                               (((value)&0x00000000ff000000ull) << 8) |
+                               (((value)&0x0000000000ff0000ull) << 24) |
+                               (((value)&0x000000000000ff00ull) << 40) |
+                               (((value)&0x00000000000000ffull) << 56));
+
+  return return_value;
+}
+#endif
+#endif
+#endif
+
+inline bigtable::bigendian64_t byteswap64(bigtable::bigendian64_t value) {
+#ifdef _MSC_VER
+  return bigtable::bigendian64_t(_byteswap_uint64(value.get()));
+#elif defined(__GNUC__) || defined(__clang__)
+  return bigtable::bigendian64_t(__builtin_bswap64(value.get()));
+#else
+  return bigtable::bigendian64_t(explicit_swap64(value.get()));
+#endif
+}
+
+constexpr char c[sizeof(int)] = {1};
+constexpr bool IsBigEndian() {
+  return c[0] != 1;  // ignore different type comparison
+}
+
+template <typename T>
```

Comment:
Please document this template: where is it used?  What is is used for?  How would an application use it to define their own encoding for their types?

Reply:
Also, it should be moved to its own file.